From de735e302987fdc8b43f2d8c294250c2bc426298 Mon Sep 17 00:00:00 2001
From: Ido Yariv <ido@wizery.com>
Date: Wed, 5 Sep 2012 22:17:09 +0300
Subject: [PATCH 12/40] wl12xx: Always pass DMA-able buffers to mmc functions

Some of the mmc drivers initiate DMA transfers with buffers passed from
higher layers. This means that the driver shouldn't ever pass non
DMA-able buffers, such as ones that are unaligned, allocated on the
stack or static.

Fix a couple of calls to the mmc layer in which buffers which weren't
necessarily DMA-able were passed.

Reported-by: "De Cesco, Jonathan" <jonathanc@ti.com>
Signed-off-by: Ido Yariv <ido@wizery.com>
---
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/boot.c   |   18 ++++++++++++------
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/cmd.c    |    6 ++----
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/event.c  |    5 ++---
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/io.h     |   10 +++++-----
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/main.c   |   31 +++++++++++++++++++++++++++++++
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/rx.c     |   11 ++++++-----
 drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/wl12xx.h |    7 +++++--
 7 files changed, 63 insertions(+), 25 deletions(-)

diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/boot.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/boot.c
index a6343c0..0651553 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/boot.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/boot.c
@@ -89,15 +89,19 @@ static void wl1271_parse_fw_ver(struct wl1271 *wl)
 
 static int wl1271_boot_fw_version(struct wl1271 *wl)
 {
-	struct wl1271_static_data static_data;
+	struct wl1271_static_data *static_data;
 	int ret;
 
-	ret = wl1271_read(wl, wl->cmd_box_addr, &static_data,
-			  sizeof(static_data), false);
+	static_data = kmalloc(sizeof(*static_data), GFP_KERNEL);
+	if (!static_data)
+		return -ENOMEM;
+
+	ret = wl1271_read(wl, wl->cmd_box_addr, static_data,
+			  sizeof(*static_data), false);
 	if (ret < 0)
-		return ret;
+		goto out;
 
-	strncpy(wl->chip.fw_ver_str, static_data.fw_version,
+	strncpy(wl->chip.fw_ver_str, static_data->fw_version,
 		sizeof(wl->chip.fw_ver_str));
 
 	/* make sure the string is NULL-terminated */
@@ -105,7 +109,9 @@ static int wl1271_boot_fw_version(struct wl1271 *wl)
 
 	wl1271_parse_fw_ver(wl);
 
-	return 0;
+out:
+	kfree(static_data);
+	return ret;
 }
 
 static int wl1271_boot_upload_firmware_chunk(struct wl1271 *wl, void *buf,
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/cmd.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/cmd.c
index 44fc971..e0f9f27 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/cmd.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/cmd.c
@@ -375,15 +375,13 @@ static int wl1271_cmd_wait_for_event_or_timeout(struct wl1271 *wl, u32 mask,
 		msleep(1);
 
 		/* read from both event fields */
-		ret = wl1271_read(wl, wl->mbox_ptr[0], &events_vector,
-				  sizeof(events_vector), false);
+		ret = wl1271_read32(wl, wl->mbox_ptr[0], &events_vector);
 		if (ret < 0)
 			return ret;
 
 		event = events_vector & mask;
 
-		ret = wl1271_read(wl, wl->mbox_ptr[1], &events_vector,
-				  sizeof(events_vector), false);
+		ret = wl1271_read32(wl, wl->mbox_ptr[1], &events_vector);
 		if (ret < 0)
 			return ret;
 
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/event.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/event.c
index be434a6..c312a3f 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/event.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/event.c
@@ -324,7 +324,6 @@ int wl1271_event_mbox_config(struct wl1271 *wl)
 
 int wl1271_event_handle(struct wl1271 *wl, u8 mbox_num)
 {
-	struct event_mailbox mbox;
 	int ret;
 
 	wl1271_debug(DEBUG_EVENT, "EVENT on mbox %d", mbox_num);
@@ -333,13 +332,13 @@ int wl1271_event_handle(struct wl1271 *wl, u8 mbox_num)
 		return -EINVAL;
 
 	/* first we read the mbox descriptor */
-	ret = wl1271_read(wl, wl->mbox_ptr[mbox_num], &mbox,
+	ret = wl1271_read(wl, wl->mbox_ptr[mbox_num], wl->mbox,
 			  sizeof(struct event_mailbox), false);
 	if (ret < 0)
 		return ret;
 
 	/* process the descriptor */
-	ret = wl1271_event_process(wl, &mbox);
+	ret = wl1271_event_process(wl, wl->mbox);
 	if (ret < 0)
 		return ret;
 
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/io.h b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/io.h
index 79b221b..8a648b3 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/io.h
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/io.h
@@ -93,13 +93,13 @@ static inline int __must_check wl1271_raw_read32(struct wl1271 *wl, int addr,
 {
 	int ret;
 
-	ret = wl1271_raw_read(wl, addr, &wl->buffer_32, sizeof(wl->buffer_32),
+	ret = wl1271_raw_read(wl, addr, wl->buffer_32, sizeof(*wl->buffer_32),
 			      false);
 	if (ret < 0)
 		goto out;
 
 	if (val)
-		*val = le32_to_cpu(wl->buffer_32);
+		*val = le32_to_cpu(*wl->buffer_32);
 
 out:
 	return ret;
@@ -108,9 +108,9 @@ out:
 static inline int __must_check wl1271_raw_write32(struct wl1271 *wl, int addr,
 						  u32 val)
 {
-	wl->buffer_32 = cpu_to_le32(val);
-	return wl1271_raw_write(wl, addr, &wl->buffer_32,
-				sizeof(wl->buffer_32), false);
+	*wl->buffer_32 = cpu_to_le32(val);
+	return wl1271_raw_write(wl, addr, wl->buffer_32,
+				sizeof(*wl->buffer_32), false);
 }
 
 /* Translated target IO */
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/main.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/main.c
index e2b99c9..b68470a 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/main.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/main.c
@@ -6309,8 +6309,36 @@ static struct ieee80211_hw *wl1271_alloc_hw(void)
 		goto err_dummy_packet;
 	}
 
+	wl->rx_mem_pool_addr = kmalloc(sizeof(*wl->rx_mem_pool_addr),
+				       GFP_KERNEL);
+	if (!wl->rx_mem_pool_addr) {
+		ret = -ENOMEM;
+		goto err_fwlog;
+	}
+
+	wl->mbox = kmalloc(sizeof(*wl->mbox), GFP_KERNEL);
+	if (!wl->mbox) {
+		ret = -ENOMEM;
+		goto err_pool_addr;
+	}
+
+	wl->buffer_32 = kmalloc(sizeof(*wl->buffer_32), GFP_KERNEL);
+	if (!wl->buffer_32) {
+		ret = -ENOMEM;
+		goto err_mbox;
+	}
+
 	return hw;
 
+err_mbox:
+	kfree(wl->mbox);
+
+err_pool_addr:
+	kfree(wl->rx_mem_pool_addr);
+
+err_fwlog:
+	free_page((unsigned long)wl->fwlog);
+
 err_dummy_packet:
 	dev_kfree_skb(wl->dummy_packet);
 
@@ -6347,6 +6375,9 @@ static int wl1271_free_hw(struct wl1271 *wl)
 	device_remove_file(wl->dev, &dev_attr_hw_pg_ver);
 
 	device_remove_file(wl->dev, &dev_attr_bt_coex_state);
+	kfree(wl->buffer_32);
+	kfree(wl->mbox);
+	kfree(wl->rx_mem_pool_addr);
 	free_page((unsigned long)wl->fwlog);
 	dev_kfree_skb(wl->dummy_packet);
 	free_pages((unsigned long)wl->aggr_buf,
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/rx.c b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/rx.c
index 1ab97cd..472643f 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/rx.c
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/rx.c
@@ -231,15 +231,16 @@ int wl12xx_rx(struct wl1271 *wl, struct wl12xx_fw_status *status)
 			mem_block = wl12xx_rx_get_mem_block(status,
 							    drv_rx_counter);
 
-			wl->rx_mem_pool_addr.addr = (mem_block << 8) +
+			wl->rx_mem_pool_addr->addr = (mem_block << 8) +
 			   le32_to_cpu(wl_mem_map->packet_memory_pool_start);
 
-			wl->rx_mem_pool_addr.addr_extra =
-				wl->rx_mem_pool_addr.addr + 4;
+			wl->rx_mem_pool_addr->addr_extra =
+				wl->rx_mem_pool_addr->addr + 4;
 
 			ret = wl1271_write(wl, WL1271_SLV_REG_DATA,
-					   &wl->rx_mem_pool_addr,
-					   sizeof(wl->rx_mem_pool_addr), false);
+					   wl->rx_mem_pool_addr,
+					   sizeof(*wl->rx_mem_pool_addr),
+					   false);
 			if (ret < 0)
 				goto out;
 		}
diff --git a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/wl12xx.h b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/wl12xx.h
index cacf638..d816bbb 100644
--- a/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/wl12xx.h
+++ b/drivers/net/wireless/compat-wireless_R5.SP2.03/drivers/net/wireless/wl12xx/wl12xx.h
@@ -440,7 +440,7 @@ struct wl1271 {
 	u32 rx_counter;
 
 	/* Rx memory pool address */
-	struct wl1271_rx_mem_pool_addr rx_mem_pool_addr;
+	struct wl1271_rx_mem_pool_addr *rx_mem_pool_addr;
 
 	/* Intermediate buffer, used for packet aggregation */
 	u8 *aggr_buf;
@@ -470,6 +470,9 @@ struct wl1271 {
 	struct delayed_work delayed_recovery;
 	bool force_mr_fw;
 
+	/* Pointer that holds DMA-friendly block for the mailbox */
+	struct event_mailbox *mbox;
+
 	/* The mbox event mask */
 	u32 event_mask;
 
@@ -500,7 +503,7 @@ struct wl1271 {
 
 	struct wl1271_stats stats;
 
-	__le32 buffer_32;
+	__le32 *buffer_32;
 	u32 buffer_cmd;
 	u32 buffer_busyword[WL1271_BUSY_WORD_CNT];
 
-- 
1.7.0.4

